### ISSUE TITLE
Why Keep <im_end> in Embeddings but Not in Labels?

### ISSUE THREAD

**Issue by VisibleCloudsDeep** [qirui-chen/MultiHop-EgoQA#4]
Dear author, thank you for your excellent work. I am a student currently studying your code. I have a question regarding the implementation in `baseline/llava/model/llava_arch.py`, specifically lines 137–146.

In the code below, why is the `<im_end>` token preserved in the embeddings (line 141), while in the labels, `cur_labels[image_token_start:image_token_start+1]` is appended (line 145), which corresponds to the `<im_start>` token label instead of the `<im_end>` token label?

```python
if getattr(self.config, 'tune_mm_mlp_adapter', False) and getattr(self.config, 'mm_use_im_start_end', False):       #137
    cur_new_input_embeds.append(self.get_model().embed_tokens(cur_input_ids[:image_token_start-1]).detach())
    cur_new_input_embeds.append(self.get_model().embed_tokens(cur_input_ids[image_token_start-1:image_token_start]))
    cur_new_input_embeds.append(cur_image_features)
    cur_new_input_embeds.append(self.get_model().embed_tokens(cur_input_ids[image_token_start+1:image_token_start+2]))
    if labels is not None:
        cur_new_labels.append(cur_labels[:image_token_start])
        cur_new_labels.append(torch.full((cur_image_features.shape[0],), IGNORE_INDEX, device=labels.device, dtype=labels.dtype))
        cur_new_labels.append(cur_labels[image_token_start:image_token_start+1])
        cur_labels = cur_labels[image_token_start+2:]                                                                                              #146
```

**Comment by qirui-chen** [comment]
Hello, thank you for your interest. I think that actually, in our model, `'mm_use_im_start_end'` is false. Therefore, there will be only one `<image>` placeholder instead of `<im_start><image><im_end>`.

https://github.com/qirui-chen/MultiHop-EgoQA/blob/95732fe219ae63378d32bbab915c077e77c17ead/baseline/llava/serve/model_worker.py#L142-L145

In other words, the following branch will be executed:

https://github.com/qirui-chen/MultiHop-EgoQA/blob/95732fe219ae63378d32bbab915c077e77c17ead/baseline/llava/model/llava_arch.py#L147-L153

**Comment by qirui-chen** [comment]
But I think your opinion is right, if `mm_use_im_start_end ` is true, the following code

https://github.com/qirui-chen/MultiHop-EgoQA/blob/95732fe219ae63378d32bbab915c077e77c17ead/baseline/llava/model/llava_arch.py#L145

should be revised as

```python
cur_new_labels.append(cur_labels[image_token_start+1:image_token_start+2])
```

**Comment by VisibleCloudsDeep** [comment]
Thank you for your  explanation
